File: app/src/main/java/deakin/gopher/guardian/view/GP/GPAddFragment.java
package deakin.gopher.guardian.view.GP;

import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.widget.ViewPager2;
import deakin.gopher.guardian.R;
import deakin.gopher.guardian.model.GP;
import deakin.gopher.guardian.util.DataListener;

public class GPAddFragment extends Fragment {
  ViewPager2 viewPager2;
  private int status;
  private GP gp;
  private DataListener dataListener;
  private EditText first_name_input,
      middle_name_input,
      last_name_input,
      clinic_address_input,
      phone_number_input,
      email_input,
      fax_input;
  private String firstName, middleName, lastName, clinicAddress, phoneNumber, email, fax;

  public GPAddFragment() {
    // Required empty public constructor
  }

  public GPAddFragment(final int status) {
    this.status = status;
  }

  @Override
  public View onCreateView(
      final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    viewPager2 = requireActivity().findViewById(R.id.dataForViewViewPager);
    final View rootView = inflater.inflate(R.layout.fragment_gp_add, container, false);
    final Button left_button = rootView.findViewById(R.id.gp_add_polygon_left);
    final Button right_button = rootView.findViewById(R.id.gp_add_polygon_right);
    final Button next_button = rootView.findViewById(R.id.gp_add_NextButton);
    final Button prev_button = rootView.findViewById(R.id.gp_add_PrevButton);
    first_name_input = rootView.findViewById(R.id.input_gp_FirstName);
    middle_name_input = rootView.findViewById(R.id.input_gp_MiddleName);
    last_name_input = rootView.findViewById(R.id.input_gp_LastName);
    clinic_address_input = rootView.findViewById(R.id.input_gp_ClinicAddress);
    phone_number_input = rootView.findViewById(R.id.input_gp_Phone);
    email_input = rootView.findViewById(R.id.input_gp_Email);
    fax_input = rootView.findViewById(R.id.input_gp_Fax);

    final Button step1_button = rootView.findViewById(R.id.step1);
    final Button step2_button = rootView.findViewById(R.id.step2);
    final Button step3_button = rootView.findViewById(R.id.step3);
    final Button step4_button = rootView.findViewById(R.id.step4);
    final Button step5_button = rootView.findViewById(R.id.step5);

    if (1 == status) {
      step4_button.setBackgroundResource(R.drawable.roundshapeseletebtn);
      step5_button.setBackgroundResource(R.drawable.roundshapebtn);
    }
    if (2 == status) {
      step5_button.setBackgroundResource(R.drawable.roundshapeseletebtn);
      step4_button.setBackgroundResource(R.drawable.roundshapebtn);
    }

    step1_button.setOnClickListener(
        view -> {
          if (dataChecker()) {
            saveGp();

            viewPager2.setCurrentItem(0, true);
          }
        });

    step2_button.setOnClickListener(
        view -> {
          if (dataChecker()) {
            saveGp();

            viewPager2.setCurrentItem(1, true);
          }
        });
    step3_button.setOnClickListener(
        view -> {
          if (dataChecker()) {
            saveGp();

            viewPager2.setCurrentItem(2, true);
          }
        });
    step4_button.setOnClickListener(
        view -> {
          if (dataChecker()) {
            saveGp();

            viewPager2.setCurrentItem(3, true);
          }
        });
    step5_button.setOnClickListener(
        view -> {
          if (dataChecker()) {
            saveGp();

            viewPager2.setCurrentItem(4, true);
          }
        });

    if (2 == status) {
      left_button.setBackgroundResource(R.drawable.polygon_3);
      right_button.setBackgroundResource(R.drawable.polygon_4);
    } else {
      next_button.setText(R.string.next);
    }

    right_button.setOnClickListener(
        view -> {
          if (1 == status) {
            // for now I keep 2 next of kins and 2 gps waiting to be add
            // but defaultly adding 1 nok and 1gp, after alick right arrow the second one shows up
            // is better
            if (dataChecker()) {
              saveGp();
              scrollPage(true);
            }
          }
        });

    left_button.setOnClickListener(
        view -> {
          if (2 == status) {
            if (dataChecker()) {
              saveGp();
              scrollPage(false);
            }
          }
        });

    next_button.setOnClickListener(
        view -> {
          // als();
          if (dataChecker()) {
            saveGp();
            if (1 == status) {
              scrollPage(true);
            } else {
              final AlertDialog.Builder builder = new AlertDialog.Builder(view.getContext());
              builder.setTitle("Saving Changes?");
              builder.setPositiveButton(
                  "YES",
                  (dialog, whichButton) -> {
                    dataListener.onDataFinished(true);
                  });
              builder.setNegativeButton("No", null);

              final AlertDialog dialog = builder.create();
              dialog.setOnShowListener(
                  arg0 -> {
                    dialog
                        .getButton(AlertDialog.BUTTON_POSITIVE)
                        .setTextColor(getResources().getColor(R.color.colorGreen));
                    dialog
                        .getButton(AlertDialog.BUTTON_NEGATIVE)
                        .setTextColor(getResources().getColor(R.color.colorRed));
                  });
              dialog.show();
            }
          }
        });

    prev_button.setOnClickListener(
        view -> {
          if (dataChecker()) {
            saveGp();
            scrollPage(false);
          }
        });
    return rootView;
  }

  private void scrollPage(final boolean isNextPage) {
    final int nextPage;
    if (isNextPage) {
      nextPage = viewPager2.getCurrentItem() + 1;
    } else {
      nextPage = viewPager2.getCurrentItem() - 1;
    }
    viewPager2.setCurrentItem(nextPage, true);
  }

  private void saveGp() {
    if (null == gp) {
      gp = new GP(firstName, middleName, lastName, phoneNumber, email, fax, null);
    } else {
      gp.setFirstName(firstName);
      gp.setMiddleName(middleName);
      gp.setLastName(lastName);
      gp.setPhone(phoneNumber);
      gp.setEmail(email);
      gp.setFax(fax);
    }
    if (1 == status) {
      dataListener.onDataFilled(null, null, null, gp, null);
    } else {
      dataListener.onDataFilled(null, null, null, null, gp);
    }
  }

  private boolean dataChecker() {
    firstName = first_name_input.getText().toString();
    middleName = middle_name_input.getText().toString();
    lastName = last_name_input.getText().toString();
    clinicAddress = clinic_address_input.getText().toString();
    phoneNumber = phone_number_input.getText().toString();
    email = email_input.getText().toString();
    fax = fax_input.getText().toString();
    if (TextUtils.isEmpty(firstName)) {
      setErrorAndReturn(first_name_input, "First name is Required.");
      return false;
    }
    if (TextUtils.isEmpty(lastName)) {
      setErrorAndReturn(last_name_input, "Last name is Required.");
      return false;
    }
    if (TextUtils.isEmpty(clinicAddress)) {
      setErrorAndReturn(clinic_address_input, "Clinic address is Required.");
    }
    if (TextUtils.isEmpty(phoneNumber)) {
      setErrorAndReturn(phone_number_input, "Phone number is Required.");
    }
    return true;
  }

  private void setErrorAndReturn(final EditText editText, final CharSequence s) {
    editText.setError(s);
  }

  @Override
  public void onAttach(@NonNull final Context context) {
    super.onAttach(context);
    try {
      dataListener = (DataListener) context;
    } catch (final ClassCastException e) {
      throw new ClassCastException(context + " must implement DataListener");
    }
  }
}

File: app/src/main/java/deakin/gopher/guardian/view/gp/GPAddFragment.kt
package deakin.gopher.guardian.view.gp

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.viewpager2.widget.ViewPager2
import deakin.gopher.guardian.R
import deakin.gopher.guardian.model.GP
import deakin.gopher.guardian.util.DataListener

class GPAddFragment(private var status: Int = 0) : Fragment() {
    private lateinit var viewPager2: ViewPager2
    private var gp: GP? = null
    private var dataListener: DataListener? = null

    private lateinit var firstNameInput: EditText
    private lateinit var middleNameInput: EditText
    private lateinit var lastNameInput: EditText
    private lateinit var clinicAddressInput: EditText
    private lateinit var phoneNumberInput: EditText
    private lateinit var emailInput: EditText
    private lateinit var faxInput: EditText

    private var firstName: String? = null
    private var middleName: String? = null
    private var lastName: String? = null
    private var clinicAddress: String? = null
    private var phoneNumber: String? = null
    private var email: String? = null
    private var fax: String? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?,
    ): View? {
        viewPager2 = requireActivity().findViewById(R.id.dataForViewViewPager)
        val rootView = inflater.inflate(R.layout.fragment_gp_add, container, false)

        val leftButton: Button = rootView.findViewById(R.id.gp_add_polygon_left)
        val rightButton: Button = rootView.findViewById(R.id.gp_add_polygon_right)
        val nextButton: Button = rootView.findViewById(R.id.gp_add_NextButton)
        val prevButton: Button = rootView.findViewById(R.id.gp_add_PrevButton)

        firstNameInput = rootView.findViewById(R.id.input_gp_FirstName)
        middleNameInput = rootView.findViewById(R.id.input_gp_MiddleName)
        lastNameInput = rootView.findViewById(R.id.input_gp_LastName)
        clinicAddressInput = rootView.findViewById(R.id.input_gp_ClinicAddress)
        phoneNumberInput = rootView.findViewById(R.id.input_gp_Phone)
        emailInput = rootView.findViewById(R.id.input_gp_Email)
        faxInput = rootView.findViewById(R.id.input_gp_Fax)

        val step1Button: Button = rootView.findViewById(R.id.step1)
        val step2Button: Button = rootView.findViewById(R.id.step2)
        val step3Button: Button = rootView.findViewById(R.id.step3)
        val step4Button: Button = rootView.findViewById(R.id.step4)
        val step5Button: Button = rootView.findViewById(R.id.step5)

        if (status == 1) {
            step4Button.setBackgroundResource(R.drawable.roundshapeseletebtn)
            step5Button.setBackgroundResource(R.drawable.roundshapebtn)
        }
        if (status == 2) {
            step5Button.setBackgroundResource(R.drawable.roundshapeseletebtn)
            step4Button.setBackgroundResource(R.drawable.roundshapebtn)
        }

        step1Button.setOnClickListener {
            if (dataChecker()) {
                saveGp()
                viewPager2.setCurrentItem(0, true)
            }
        }

        step2Button.setOnClickListener {
            if (dataChecker()) {
                saveGp()
                viewPager2.setCurrentItem(1, true)
            }
        }

        step3Button.setOnClickListener {
            if (dataChecker()) {
                saveGp()
                viewPager2.setCurrentItem(2, true)
            }
        }

        step4Button.setOnClickListener {
            if (dataChecker()) {
                saveGp()
                viewPager2.setCurrentItem(3, true)
            }
        }

        step5Button.setOnClickListener {
            if (dataChecker()) {
                saveGp()
                viewPager2.setCurrentItem(4, true)
            }
        }

        if (status == 2) {
            leftButton.setBackgroundResource(R.drawable.polygon_3)
            rightButton.setBackgroundResource(R.drawable.polygon_4)
        } else {
            nextButton.setText(R.string.next)
        }

        rightButton.setOnClickListener {
            if (status == 1) {
                if (dataChecker()) {
                    saveGp()
                    scrollPage(true)
                }
            }
        }

        leftButton.setOnClickListener {
            if (status == 2) {
                if (dataChecker()) {
                    saveGp()
                    scrollPage(false)
                }
            }
        }

        nextButton.setOnClickListener {
            if (dataChecker()) {
                saveGp()
                if (status == 1) {
                    scrollPage(true)
                } else {
                    val builder = AlertDialog.Builder(it.context)
                    builder.setTitle("Saving Changes?")
                    builder.setPositiveButton("YES") { _, _ ->
                        dataListener?.onDataFinished(true)
                    }
                    builder.setNegativeButton("No", null)

                    val dialog = builder.create()
                    dialog.setOnShowListener {
                        dialog.getButton(AlertDialog.BUTTON_POSITIVE)
                            .setTextColor(resources.getColor(R.color.colorGreen))
                        dialog.getButton(AlertDialog.BUTTON_NEGATIVE)
                            .setTextColor(resources.getColor(R.color.colorRed))
                    }
                    dialog.show()
                }
            }
        }

        prevButton.setOnClickListener {
            if (dataChecker()) {
                saveGp()
                scrollPage(false)
            }
        }

        return rootView
    }

    private fun scrollPage(isNextPage: Boolean) {
        val nextPage = if (isNextPage) viewPager2.currentItem + 1 else viewPager2.currentItem - 1
        viewPager2.setCurrentItem(nextPage, true)
    }

    private fun saveGp() {
        if (gp == null) {
            gp = GP(
                firstName ?: "",
                middleName ?: "",
                lastName ?: "",
                phoneNumber ?: "",
                email ?: "",
                fax ?: "",
                null.toString()
            )
        } else {
            gp?.apply {
                firstName = this@GPAddFragment.firstName ?: ""
                middleName = this@GPAddFragment.middleName ?: ""
                lastName = this@GPAddFragment.lastName ?: ""
                phone = this@GPAddFragment.phoneNumber ?: ""
                email = this@GPAddFragment.email ?: ""
                fax = this@GPAddFragment.fax ?: ""
            }
        }
        if (status == 1) {
            dataListener?.onDataFilled(null, null, null, gp, null)
        } else {
            dataListener?.onDataFilled(null, null, null, null, gp)
        }
    }

    private fun dataChecker(): Boolean {
        firstName = firstNameInput.text.toString()
        middleName = middleNameInput.text.toString()
        lastName = lastNameInput.text.toString()
        clinicAddress = clinicAddressInput.text.toString()
        phoneNumber = phoneNumberInput.text.toString()
        email = emailInput.text.toString()
        fax = faxInput.text.toString()

        return when {
            firstName.isNullOrEmpty() -> {
                setErrorAndReturn(firstNameInput, "First name is Required.")
                false
            }

            lastName.isNullOrEmpty() -> {
                setErrorAndReturn(lastNameInput, "Last name is Required.")
                false
            }

            clinicAddress.isNullOrEmpty() -> {
                setErrorAndReturn(clinicAddressInput, "Clinic address is Required.")
                false
            }

            phoneNumber.isNullOrEmpty() -> {
                setErrorAndReturn(phoneNumberInput, "Phone number is Required.")
                false
            }

            else -> true
        }
    }

    private fun setErrorAndReturn(editText: EditText, message: CharSequence) {
        editText.error = message
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        dataListener = try {
            context as DataListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement DataListener")
        }
    }
}

